### PR TITLE
Jira integration

### PR DESCRIPTION
**Pull request by gmatthews20**
### Description:

- API for creating a JIRA issue in a given board
- Test workflow for creating issue
- Templating for JIRA issue descriptions

---

### Submitter:

Have you (where applicable):

* [x] Added unit tests?
* [x] Checked the latest commit runs on Dev?
* [ ] Updated the example config file(s) and README?

---

### Reviewer

Does this PR:

* [ ] Place non-StackStorm code into the `lib` directory?
* [ ] Have unit tests for the action/sensor and `lib` layers?
* [ ] Have clear and obvious action parameter names and descriptions?

### CODE REVIEW

File: lib/structs/jira/jira_account.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        Returns instance of this dataclass from a dictionary (for loading from config)
+        """
+        field_set = {field.name for field in fields(JiraAccount) if field.init}
+        filtered_arg_dict = {
+            key: value for key, value in dictionary.items() if key in field_set
+        }
+        return JiraAccount(**filtered_arg_dict)
+
+    @staticmethod
+    def from_pack_config(pack_config: dict, jira_account_name: str):
```
Comment (anish-mudaraddi) [inline lib/structs/jira/jira_account.py:30]: you are missing tests for this function and `from_dict` function - see  https://github.com/stfc/st2-cloud-pack/blob/main/tests/structs/test_smtp_account.py for similar tests

File: lib/jira_api/jira_template_schemas.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# This file holds configuration information on each template
```
Comment (anish-mudaraddi) [inline lib/jira_api/jira_template_schemas.yaml:1]: Something for a future PR - now that there's two template directories - we should merge them into one 
I'd really like there to be a `templates` directory at the top, then a jira/email subdirectory - each with equivalent schema yaml file and template subdirectories underneath

**Review by anish-mudaraddi — CHANGES_REQUESTED** [review]

**Review by anish-mudaraddi — DISMISSED** [review]

File: actions/src/openstack_actions.py [hunk 3]
```diff
@@ -42,4 +43,9 @@ def parse_configs(self, **kwargs):
                 self.config, kwargs["smtp_account_name"]
             )
             del kwargs["smtp_account_name"]
+        if "atlassian_account_name" in kwargs:
+            kwargs["atlassian_account"] = JiraAccount.from_pack_config(
+                self.config, kwargs["atlassian_account_name"]
+            )
+            del kwargs["atlassian_account_name"]
```
Comment (meoflynn) [inline actions/src/openstack_actions.py:50]: Why does the `atlassian_account_name` have to be removed from `kwargs` if it is found in kwargs?

File: lib/jira_api/jira_issue.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.conn.close()
+
+
+def create_jira_task(account, issue_details: IssueDetails) -> str:
+    """
+    Creates a JIRA issue in a given project
+    :param project_id: ID or key of the JIRA project
+    """
```
Comment (meoflynn) [inline lib/jira_api/jira_issue.py:36]: Could you add what is returned by this method in the docstring?

File: lib/jira_api/template_handler.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            raise JiraTemplateError(
+                f"Missing attributes required for template {template_details.template_name}: {', '.join(missing_attrs)}"
+            )
+
+        return attrs
+
+    def _get_template_file(self, template_fp) -> Template:
+        """
+        Method to read in template file as a jinja template from filepath.
+        :param template_fp: a relative filepath for template file starting from 'JIRA_TEMPLATE_ROOT_DIR'
```
Comment (meoflynn) [inline lib/jira_api/template_handler.py:89]: Could you add what this method returns in the docstring - what kind of `Template` is being returned?

File: lib/jira_api/template_handler.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            template_metadata
+            if template_metadata
+            else self._load_all_metadata(self.JIRA_TEMPLATE_METADATA_FP)
+        )
+
+    @staticmethod
+    def _load_all_metadata(metadata_fp) -> Dict:
+        """
+        Method which reads in jira templates yaml file. This file holds metadata for each template name.
+        """
```
Comment (meoflynn) [inline lib/jira_api/template_handler.py:47]: Could you add what values in the dictionary are being returned by the method?

File: lib/jira_api/template_handler.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        except TemplateNotFound as not_found_exp:
+            raise JiraTemplateError(
+                "Could not find template with filepath {self.JIRA_TEMPLATE_ROOT_DIR / template_fp}"
+            ) from not_found_exp
+
+    def _get_template_metadata(self, template_name) -> Dict:
+        """
+        Method to get metadata values for a specific template from all metadata file
+        :param template_name: name of template to get metadata for
+        """
```
Comment (meoflynn) [inline lib/jira_api/template_handler.py:102]: Could you add what values in the dictionary are being returned by the method?

File: lib/jira_api/template_handler.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+    def _render_template(
+        self,
+        template_details: JiraTemplateDetails,
+        file_path_key: str,
+    ) -> str:
+        """
+        Helper method to render template
+        :param template_details: dataclass representing user-defined values for template
+        :param file_path_key: metadata dictionary key which holds the filepath pointing to template file
```
Comment (meoflynn) [inline lib/jira_api/template_handler.py:130]: What string is returned by this method?

**Review by meoflynn — CHANGES_REQUESTED** [review]

File: actions/src/openstack_actions.py (hunk 3 above)
Comment (gmatthews20) [reply actions/src/openstack_actions.py:50]: `atlassian_account_name` is just used to load the correct token and username from the stackstorm config which are then used by the action, `atlassian_account_name` is not needed to connect to jira

File: lib/jira_api/template_handler.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+    @staticmethod
+    def _parse_template_attrs(
+        template_details: JiraTemplateDetails, template_schema: Dict[str, str]
+    ) -> Dict[str, str]:
+        """
+        Static method which returns values for all attributes to replace for the template that has been provided.
+        attribute values either set to mapped values in 'given_vals' or set to a default given by 'template_schema'
+        :param template_details: A dataclass containing info on templates
+        :param template_schema: A dictionary holding required attribute names mapped to defaults
```
Comment (meoflynn) [inline lib/jira_api/template_handler.py:65]: Could we add what is returned in this method?

**Review by meoflynn — CHANGES_REQUESTED** [review]

File: actions/src/openstack_actions.py (hunk 3 above)
Comment (meoflynn) [reply actions/src/openstack_actions.py:50]: Could you add a comment for this step in the method?

**Review by meoflynn — APPROVED** [review]

**Review by anish-mudaraddi — APPROVED** [review]

**Pull request merged**